index.bs

```diff
@@ -3844,19 +3850,35 @@ The [=remote end steps=] with |command parameters| are:
    1. Otherwise, set the |context|'s [=layout viewport=] to the
       implementation-defined default.
 
+1. Run the [[cssom-view-1#resizing-viewports]] steps.
+
 1. If |command parameters| [=map/contains=] the <code>devicePixelRatio</code> field:
 
    1. Let |device pixel ratio| be the |command
       parameters|["<code>devicePixelRatio</code>"].
 
-   1. If |device pixel ratio| is not null, change the size of the [=CSS Pixel=]
-      in |context|'s [=layout viewport=] such that it corresponds to |device
-      pixel ratio| in device pixels.
+   1. For the |context| and all descendant traversables:
 
-   1. Otherwise, set the size of the [=CSS Pixel=] in |context|'s [=layout
-      viewport=] to the implementation-defined default.
+      1. If |device pixel ratio| is not null:
 
-1. Run the [[cssom-view-1#resizing-viewports]] steps.
+         1. When the [=select an image source from a source set=] are run, act as if
+            the implementation's pixel density was set to |device pixel ratio| when selecting an image.
+
+         1. For the purposes of the [=resolution media feature=], act as if
+            the implementation's resolution is |device pixel ratio| dppx scaled by the page zoom.
+
+         1. Run [=evaluate media queries and report changes=] for [=document=] currently loaded
+            in a specified |context|.
+
+      1. Otherwise:
+
+         1. Reset special behavior for the [=select an image source from a source set=],
```

Comment:
I don't love the "Reset special behavior" formulation; it's more or less clear in context what it must mean, but "special behaviour" isn't defined. Maybe we can say something more like "`When the [=select an image source from a source set=] steps are run, use the implementation's default behavior, without any changes made by previous invocations of these steps.` or something like that. It's still not really clear, but short of actually defining how this stuff works, it might be the best we can do. Similarly below.

Reply:
Yeah, I also didn't like this so much, but didn't come up with anything better.
Updated now, as you suggested.